### ISSUE TITLE
Modifier classes

### ISSUE THREAD

**Issue by ai** [callstack/linaria#234]
**Do you want to request a *feature* or report a *bug*?**

feature

**What is the expected behavior?**

To have the same modifiers syntax as in [`astroturf`](https://github.com/4Catalyzer/astroturf)

```js
const Button = styled('button')`
  color: black;
  border: 1px solid black;
  background-color: white;

  &.primary {
    color: blue;
    border: 1px solid blue;
  }

  &.color-green {
    color: green;
  }
`;

<Button primary color="green">
```

Compare to `${  }` interpolation it keeps CSS syntax. CSS highlight and autocomplete works.

**Comment by satya164** [comment]
@thymikee @zamotany thoughts?

**Comment by thymikee** [comment]
I think it introduces a level of indirection and complicates stuff. I'm not a fan.

**Comment by ai** [comment]
@thymikee is it OK to simplify a lot of user’s code but increase complexity in a framework. Is how all features ware made?

Why you are not a fan of it?

**Comment by ai** [comment]
> What about prop name clashes if we wrap custom components with styled()?

Can you show a small example? I am thinking that I was worry about it too, but not sure that we are speaking about the same.

**Comment by satya164** [comment]
For example, say:

```js
const Wrapper = styled(MyComponent)`
  &.primary {
    color: blue;
    border: 1px solid blue;
  }
`;

<Wrapper primary />
```

The `primary` prop is also going to be passed to `MyComponent` unless we filter it out. But we don't know supported props, so not possible filter it out.

Wondering something like this is a good idea:

```js
<Wrapper {...mx({ primary: true, color: 'green' })} />
```

Or maybe the user can use this package: https://npmjs.com/package/classnames

I'm not opposed to the idea, but probably need to think a bit more to avoid the edge cases.

**Comment by thymikee** [comment]
Was about to write the same, but Satya 🤖 was quicker!

**Comment by ai** [comment]
@satya164 why we don’t know about supported props?

We have `styled` object from CSS Modules and can look into classes. `astroturf` does the same to detect what properties should be used as style modifiers. https://github.com/4Catalyzer/astroturf/blob/master/src/runtime/styled.js#L23-L25

**Comment by satya164** [comment]
@ai I'm a bit confused about this line:

```js
const hasModifiers = Object.keys(styles).some(
   className => className !== camelName && className !== kebabName,
);
```

Is `styles` the list of `styles` applied to the component and astroturf keeps them in the bundle?

BTW I was not talking about which props are used in the CSS. I was talking about the props that are accepted by `MyComponent` (in above case). We don't know if a prop is used only for styling or is also a regular prop. If the user uses a prop only for styling and later someone else adds that prop to the component without checking the callsites, it might change the behaviour.

It's also a problem with property interpolation though, so not specific to this approach. I'm thinking if there is a way to distinguish style only props. Maybe instead of `primary`, the user can write `$primary`, and we know for sure that it's a prop only used for styling, and we filter them out.

**Comment by ai** [comment]
@satya164 

> Is styles the list of styles applied to the component and astroturf keeps them in the bundle?

Yeap, `styles` is that object from CSS Modules `{ list: 'component_list_hb4h5h45' }`. We need it anyway when we use `styled()`.

> I was talking about the props that are accepted by MyComponent (in above case). 

We can have the agreement that if prop is defined by class in `styled()`, it will not be passed down. It is pretty expected behaviour. If you declare prop as a class in styles, it definitely stops to be a just regular prop.

**Comment by satya164** [comment]
Also, what do you think about `data-` attributes? This is already supported:

```
const Box = styled.div`
  /* some css */

  &[data-primary] {
    border: 2px solid blue;
  }  

  &[data-color=green] {
    color: green;
  }
`;

<Box data-primary data-color="green" />
```

**Comment by ai** [comment]
1. `data-` attributes will be DOM
2. It requires more symbols
3. It is also a little hacky. `data-` attributes were created for a different purpose. People will not love them, as result it will not be syntax sugar feature for linaria as it works for `astroturf`.

**Comment by satya164** [comment]
Makes sense. We'll think about it a bit more then.

Any reason you don't like following:

```js
<Box $primary $color="green" />
```

One extra character, but I prefer to be explicit rather than assuming which properties shouldn't be passed.

**Comment by satya164** [comment]
@ai currently linaria doesn't use `css-modules`, just normal `css-loader`, so we don't have a `styles` object. it can be implemented though. though the reason I prefer `$primary` is because it's explicit that it's for styling, and not a regular prop/attribute.

**Comment by ai** [comment]
> currently linaria doesn't use css-modules

How do you scope keyframes? Or what if I don’t want to scope class (`:global()` in CSS or `injectGlobal` in SC)?

Did I understand correctly that right now you collect all styles to the single file `styles.css`? How code splitting works?

**Comment by satya164** [comment]
> How do you scope keyframes? Or what if I don’t want to scope class

Currently, we use `stylis` as the parser (though planning to switch to postcss for few reasons) which handles scoped keyframes as well as `:global`.

> Did I understand correctly that right now you collect all styles to the single file styles.css? How code splitting works?

We create separate CSS files for each JS file and add a require statement, which gets picked up by style loader. Code splitting works similar to when you import vanilla css files with `mini-css-extract-plugin`.

**Comment by brandonkal** [comment]
Here's an idea if you want to keep nesting rather than composing individual classnames that would work with linaria's styled approach of using CSS variables for interpolations:

```js
const Button = styled.button`
  color: black;
  border: 1px solid black;
  background-color: white;

  &[props|primary] {
    color: blue;
    border: 1px solid blue;
  }

  &[props|color=green] {
    color: green;
  }
`;

<Button primary color="green">
```

Linaria would then transform it to this before running stylis:

```js
const Button = styled.button`
  color: black;
  border: 1px solid black;
  background-color: white;

  &[style*="--a"] {
    color: blue;
    border: 1px solid blue;
  }

  &[style*="--b"] {
    color: green;
  }
`;

<Button primary color="green">
```
Linaria would then accept the props for primary and color and if they match what was specified, it would apply the appropriate styles:

```html
<button class="b13mnax5" style="--a:1;">Primary Button</button>
```
Here the value 1 is not important, we are matching for the word "--a".
This feels a little hacky. These really could be converted to classnames instead, applied based on state, but I am leaving it as an option to consider.

What I like: Using css namespaces rather than JS interpolation. It is small annoyance to type out a function every time: `${props => props.primary || 'fallback'}`

**Comment by brandonkal** [comment]
I've added a [comment](https://github.com/callstack/linaria/issues/427#issuecomment-497515487) to my RFC that specifically addresses this issue. I've included a lot of detail, but the essence of it is this:
Rather than using an arbitrary prefix for style only props, instead specify in the CSS rule if a prop should not be forwarded.
Proposed syntax is like this:
```css
&[props|primary--] {}
```
`--` suffix, do not pass prop to children. Props are passed to children by default (unless it is a styled DOM node where valid props are known). Simply include a `--` to opt out of this for a specific prop.

**Comment by ShanonJackson** [comment]
Not a fan of the syntax especially in terms of static typing.

Maybe something like this...

```
styled.button`
      color: red;
`.variants({
       primary: css`color: blue`,
       secondary: css`color: white`
});
```

**Comment by brandonkal** [comment]
Yes, I have revised by thoughts on the syntax. This is what [I am using](https://www.npmjs.com/package/@brandonkal/linaria) for my ~60 component UI library and it works out quite well. It feels a lot closer to CSS syntax IMO and works with static typing:
```js
interface ButtonProps {
  primary$?: boolean;
}
 
const Button = (p => styled.button<ButtonProps>`
  background: black;
  color: white;
  &${[p.primary$]} {
    background: blue;
    color: black;
    &:hover {
      border-color: red;
    }
  }
`)({} as ButtonProps);
```

This works much better for static typing than my comment above.

The only issue I have run into with my syntax is I cannot target the primary modifier in another component. This isn't so bad, as styles are self contained, but I did find I wanted to do this once.

Your approach is interesting, but I am not a fan of having to call the css function everywhere. Also, accepting an array feels very much like reimplementing classnames and you lose out on accepting an arbitrary prop that can be used for other logic (you'd have to do let isPrimary = props.variant.includes('primary') in your render functions rather than destructuring a prop).

Another downside to your approach is you can't target multiple variants, i.e. applying some styles when both primary and disabled are enabled.

**Comment by ShanonJackson** [comment]
There's an example there with multiple modifiers

**Comment by brandonkal** [comment]
No there is not. You misunderstood what I intended to convey.
Take this css:
```css
.button {
  background: pink;
}
.button__primary {
  background: green;
}
.button__disabled {
  cursor: default;
}
/* This */
.button__disabled.button__primary {
  color: red;
}
```

**Comment by ShanonJackson** [comment]
Thought it about this for awhile the solution would be this:

styled.div generates an object (not a string like css), therefore just add modifiers as statics on that object.

I.E
```
const Button = styled.button`
      color: red;
`.variants({
       primary: css`color: blue`,
       secondary: css`color: white`,
       disabled: css`cursor: default`
});


const TargetDisabledButton = css`
    .${Button.disabled} {
           color: coralpink;
    }
`
```

**Comment by ShanonJackson** [comment]
I thought for along time about how to word this, in a way that is in constructive but also critical of how this library is being maintained.

This library has so much potential, infact i migrated all our stuff (1000+ components) off styled-components to linaria because i believed it made better trade offs especially around performance and server-side rendering.

So the question is this, have the staff at callstack just simply stopped giving a fuck about this library? why have pull requests sat there for months? why is the typescript support still terrible? why is there no support for modifiers? why are people having to fork this library to make common sense changes (I.E @brandonkal ). Why is it still using an ancient version of core-js leading to many problems in many build pipelines. Why are there just no improvements?

I'm a huge supporter of Linaria company wide, and on my personal projects. Please don't just waste this library's potential its an amazing library and in my opinion the best css-in-js library currently out there in terms of its approach.

**Comment by brandonkal** [comment]
That is a creative solution but I am still not a fan. To provide some reasons:
- It necessitates a lot of extra boilerplate and feels less like writing CSS.
For instance your example would require always importing css and writing it for every variation:
```js
import { css } from 'linaria'
const Button = styled.button`
      color: red;
`.variants({
       primary: css`
         color: blue
       `,
       secondary: css`
         color: white
       `,
       disabled: css`
         cursor: default
      `
});
```
Even for short strings, Prettier is typically going to break it into more lines of code.
- The variants array feels like a pointless re-implementation of cx. Writing code like this:
```js
<Card variants={[disabled && "disabled"]}/>
// rather than
<Card disabled$ />
```
Doesn't seem like a win. You end up typing everything twice.
- Any logic cannot be contained in the styled component.
An example of something that isn't possible with the last suggestion:

```js
interface ButtonProps {
  primary$?: boolean;
}
 
const Button = styled.button<ButtonProps>`
  color: white;
  &${[window.isUserLoggedIn]} {
    background: blue;
  }
`)({} as ButtonProps);

That logic could of course be more complex than checking the truthiness of a window variable. The last suggestion would require repeating that logic everywhere Button is used.

I believe there should be a way to add static typing (i.e. accessing modifiers as if they were properties of a styled component (they are, but only after the linaria transform) to my suggested approach.

**Comment by ShanonJackson** [comment]
- let's change the implementation of variants to not need css keyword just a literal is fine (under the hood it will use css
- let's change the variants array to your syntax with just the modifier name on the component.
- that last example is possible you just wouldn't do it like that you'd call the modifier name "isLoggedIn" and then apply it in the component on the state window.isUserLoggedIn, this separates the concerns of styles and state.

**Comment by brandonkal** [comment]
The reason I say the object approach requires the css keyword is because no syntax highlighter or linter will accept a plain literal.
That last example is just to illustrate that you can link state to style logic with the approach I suggested. The logic could be much more complex than just reading something off window in my simple example. The approach is inspired by LinkedIn's CSSBlocks if you wish to look into that.

**Comment by Jayphen** [comment]
>Currently, we use stylis as the parser (though planning to switch to postcss for few reasons)

@satya164 What were those reasons?

I disabled stylis in my build, but haven't been able to get global styles working properly without it yet.

**Comment by ShanonJackson** [comment]
Linters and and code highlighters are just tooling; I'd rather rewrite the tooling based on ideal implementation than use a non-ideal implementation and avoid haivng to rewrite the tooling.

Syntax highlighting and linters will 100% accept a plain literal you just wont trigger it automatically on any literal you'll look to see if that literal looks like it belongs to styled component; Walk the AST.

**Comment by alexojegu** [comment]
I have few experience with CSS and with Linaria, but this seems to work:

```ts
interface TestProps {
    $color: string;
    className?: string;
}

const Test = styled.a<TestProps>`
    text-decoration: none;
    &.foo {
        color: ${({ $color }) => $color};
    }
    &.bar {
        font-weight: 900;
    }
`;

<Test href="#" $color="red" className="foo">Text</Test>
<Test href="#" $color="blue" className={cx("foo", "bar")}>Text</Test>
<Test href="#" $color="red"  className={cx( "foo", "bar",  css`
    background-color: blue;
` )}>Text</Test>
```

And this too:
```ts
const Test2 = styled(Test)`
    &.bar {
        font-size: 32px;
    }
`;

<Test2 href="#" $color="red"  className={cx( "foo", "bar", css`
    background-color: blue;
` )}>Text</Test2>
```

Yeah, the DX with static typing is bad.

**Comment by dpashk-figma** [comment]
Just want to add my vote to the original proposal, I'll copy-paste it to reduce the up and down scrolling:
```jsx
const Button = styled('button')`
  color: black;
  border: 1px solid black;
  background-color: white;

  &.primary {
    color: blue;
    border: 1px solid blue;
  }

  &.color-green {
    color: green;
  }
`;

<Button primary color="green">
```

I used astroturf extensively and was a big fan of the auto-generated modifier class names as illustrated in the original post. It's much more elegant and idiomatic than having a bunch of lines that look like this `color: ${props => (props.primary ? 'tomato' : 'black')};`. Besides the latter syntax looking  bulky, it also doesn't let you syntactically group all the style rules that you want to apply if the "primary" modifier is set. 

The OP also requires zero boilerplate vs the `.variants({})` proposal above, for instance. You just write your styles in pretty much the same way you used to CSS stylesheets with semantic class names back in the day, except these semantic class names are set from component props. It really makes a difference in the developer experience.

**Comment by dreyks** [comment]
btw, I've made a "pre-loader" for linaria that makes it behave like astroturf by replacing modifier classes with `css` templates

basically it turns this
```js
import { styled } from '@linaria/react'

export const Container = styled('div')`
  background-color: rebeccapurple;

  &.something {
    width: 100%;
  }
`
```

into this
```js
import { styled } from '@linaria/react'
import { css } from '@linaria/core'
import { getRuntimeNode } from '../../../runtime.js'


const linariaAutopilotInterpolations0 = {
  'something': css``
}

const LinariaAutopilotInternalComponent0 = styled('div')`
  background-color: rebeccapurple;

  &.${linariaAutopilotInterpolations0['something']} {
    width: 100%;
  }
`

export const Container = getRuntimeNode(LinariaAutopilotInternalComponent0, linariaAutopilotInterpolations0)
```

the getRuntimeNode is similar to what astroturf has

it works good, but it has some flaws because it meant to be a fast-and-dirty implementation, so its cuts some corners and I sadly don't have time atm to make it right. I'll try to check out wyw-in-js, maybe it'll be easier

**Comment by dpashk-figma** [comment]
@dreyks are you able to share your pre-loader code?

**Comment by dreyks** [comment]
yup, here it is https://gist.github.com/dreyks/36ffcd4734b56b8f0c32a78127925d83. can be used as a webpack loader, jest transform, or manually from a script.
the main issue right now is that components created this way can't be re-styled as `styled(Something)` because the actual variable name (Something) affects linaria classname generation, and after going through the preloader the name gets changed to LinariaAutoPilotInternalComponent

**Comment by espretto** [comment]
Hello,

I'd like to bring @satya164 's [`data-*` attributes](https://github.com/callstack/linaria/issues/234#issuecomment-426431752) back to the table and comment on @ai 's points:

> 1. `data-` attributes will be DOM

Modifier class names will also end up in the DOM. (This is actually good for testing)

>     2. It requires more symbols

Yes it does. I can't argue with that. But with type-safety comes auto-completion (read on).

>     4. It is also a little hacky. `data-` attributes were created for a different purpose. People will not love them, as result it will not be syntax sugar feature for linaria as it works for `astroturf`.

I am afraid I disagree. Using `data-*` attributes is definitely an improvement over template string interpolations. Also, it is valid to hold state in attributes, HTML itself does it (e.g. `checked`, `disabled` etc.) so I wouldn't say `data-*` attributes were misused to hold custom state. I'd even say the approach is widely accepted (see [mantine](https://mantine.dev/styles/data-attributes/)).

Another advantage of `data-*` attributes is that they don't have to be added to and removed from the css class list or be filtered out of properties spread onto jsx elements (the whole reason non-standard property prefix `$` was introduced by [styled-components](https://styled-components.com/docs/basics#passed-props)). They work out of the box and can be made type-safe like this:

```tsx
import { modifier } from "modifier";

interface ComponentProps {
  "data-variant": "light" | "dark"
}

export const Component = styled.div<ComponentProps>`
  [...]

  ${modifier("data-variant", "light")} {
    color: gray;
  }

  ${modifier("data-variant", "dark")} {
    color: white;
  }
`;
```
**modifier.ts**
```ts
type primitive = string | number | boolean;
type reserved = "data-test-id";

export function modifier<
  T extends object,
  U extends Exclude<keyof T, symbol | reserved>,
  V extends T[U] & primitive,
>(prop: U, value: V): ((props: T) => string) | string {
  return `&[${prop}="${value}"]`;
}
```

**Comment by dreyks** [comment]
I think the argument on hackiness comes from the fact that what we're discussing here is how the component looks (styling) and data-attributes are designed for holding, well, data. 
One can argue that data indeed does define how the component looks, but the main discussion and the holy war here is that it should do so through the intended instruments, i.e., classnames
On the other hand, you're correct when you say that storing state in DOM is already a thing, so there's that

Also, another advantage of the classname modifiers is that they can be hashed/obfuscated, while data-* attributes will be spelled out fully which might be quite a lot for a component with many variants

**Comment by espretto** [comment]
> it should do so through the intended instruments, i.e., classnames

Attribute selectors are not less valid CSS. They even have the same specificity as classnames. HTML itself represents state (=data) through attributes and it is no hack to use attribute selectors to target them:
```css
[disabled] {
  color: gray;
}
```

I assume byte-size is a concern, then. Obfuscation/hashing will surely improve it but (e2e-)testing and debugging will be more difficult.

**Comment by ai** [comment]
I moved to Svelte, so you can close it if `[data-*]` is the default API for that (I little more complex without type checking, but works too)
